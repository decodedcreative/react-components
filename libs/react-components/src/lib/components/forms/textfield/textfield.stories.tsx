import { ComponentStory, ComponentMeta } from '@storybook/react';
import { TextField } from './textfield';

export default {
  title: 'Theme/Forms/TextField',
  component: TextField,
  parameters: {
    backgrounds: {
      default: 'Light',
      values: [
        { name: 'Light', value: '#ffffff' },
        { name: 'Dark', value: '#161b2f' },
      ],
    },
  },
  argTypes: {},
  args: {},
} as ComponentMeta<typeof TextField>;

const Template: ComponentStory<typeof TextField> = (args) => (
  <TextField {...args} />
);

export const Default = Template.bind({});
Default.args = {};

export const Error = Template.bind({});
Error.args = {
  'aria-invalid': true,
  defaultValue: 'Error',
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
  defaultValue: 'Disabled',
};
