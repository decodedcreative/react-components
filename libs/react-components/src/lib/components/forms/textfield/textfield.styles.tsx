import styled, { css } from 'styled-components';
import { TextFieldProps } from './textfield';

export const TextField = styled('div')<TextFieldProps>`
  width: 100%;
`;

export const TextFieldInput = styled('input')`
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: ${(props) => props.theme.colors.color70};
  font-family: ${(props) => props.theme.typography.fontFamily};
  font-size: ${(props) => props.theme.typography.sizes.default};
  line-height: ${(props) => props.theme.typography.lineHeight};
  border: ${(props) => props.theme.general.borderWidth} solid
    ${(props) => props.theme.components.forms.borderColor};

  &:focus::not(&:focus-visible) {
    outline: 0;
    border-color: unset;
  }

  &:focus-visible {
    color: ${(props) => props.theme.colors.color60};
    border-color: ${(props) => props.theme.colors.interactive};
    outline: ${(props) => props.theme.general.focusWidth} solid transparent;
    outline-offset: ${(props) => props.theme.general.focusWidth};
    box-shadow: ${(props) => props.theme.general.focusBoxShadow};
  }

  &:disabled {
    color: ${(props) => props.theme.colors.color50};
    cursor: not-allowed;
  }

  ${(props) =>
    props['aria-invalid'] &&
    css`
      border-color: ${(props) => props.theme.colors.error};
      color: ${(props) => props.theme.colors.error};
    `}
`;
