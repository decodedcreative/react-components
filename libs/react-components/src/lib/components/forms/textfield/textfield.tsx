import { ComponentPropsWithoutRef, forwardRef } from 'react';
import * as S from './textfield.styles';

export interface TextFieldProps extends ComponentPropsWithoutRef<'input'> {
  disabled?: boolean;
  wrapperProps?: ComponentPropsWithoutRef<'div'>;
}

export const TextField = forwardRef<HTMLInputElement, TextFieldProps>(
  (
    {
      'aria-invalid': ariaInvalid,
      required,
      type = 'text',
      wrapperProps,
      ...props
    },
    ref
  ) => (
    <S.TextField {...wrapperProps}>
      <S.TextFieldInput
        aria-invalid={ariaInvalid}
        aria-required={required}
        ref={ref}
        type={type}
        {...props}
      />
    </S.TextField>
  )
);
