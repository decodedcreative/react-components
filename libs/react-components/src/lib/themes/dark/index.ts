import colors from './colors';
import components from './components';
import general from './general';
import spacing from './spacing';
import typography from './typography';

const darkTheme = {
  name: 'darkTheme',
  colors,
  components,
  general,
  spacing,
  typography,
};

export default darkTheme;
