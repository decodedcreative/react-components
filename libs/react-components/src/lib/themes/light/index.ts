import colors from './colors';
import components from './components';
import general from './general';
import typography from './typography';

const lightTheme = {
  name: 'lightTheme',
  colors,
  components,
  general,
  typography,
};

export default lightTheme;
