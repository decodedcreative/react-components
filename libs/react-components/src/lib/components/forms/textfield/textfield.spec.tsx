import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { lightTheme } from '../../../themes';
import { TextField, TextFieldProps } from './textfield';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

const WithTheme = ({ ...props }) => (
  <ThemeProvider theme={lightTheme}>
    <TextField {...props} />
  </ThemeProvider>
);

const createComponentJSX = ({ ...props }: TextFieldProps) => (
  <WithTheme {...props} />
);
const createComponentRender = ({ ...props }: TextFieldProps) =>
  render(<WithTheme {...props} />);

describe('Textfield', () => {
  it('should render successfully', () => {
    const component = createComponentRender({} as TextFieldProps);
    expect(component).toBeTruthy();
  });

  it('should match the default snapshot', () => {
    const defaultTextField = createComponentJSX({} as TextFieldProps);
    const defaultTextFieldJSON = renderer.create(defaultTextField).toJSON();
    expect(defaultTextFieldJSON).toMatchSnapshot();
  });

  describe('TextField - States', () => {
    describe('TextField - States - Error', () => {
      it('should have an aria-invalid attribute on its input if in an error state', () => {
        const component = createComponentRender({
          'aria-invalid': true,
          'data-testid': 'textfield-input',
        } as TextFieldProps);
        const input = component.getByTestId('textfield-input');
        expect(input).toHaveAttribute('aria-invalid');
      });
      it('should match the TextField With Error snapshot', () => {
        const errorTextField = createComponentJSX({
          'aria-invalid': true,
        } as TextFieldProps);
        const errorTextFieldJSON = renderer.create(errorTextField).toJSON();
        expect(errorTextFieldJSON).toMatchSnapshot();
      });
    });
    describe('TextField - States - Disabled', () => {
      it('should disable its input if a disabled prop is supplied', () => {
        const component = createComponentRender({
          disabled: true,
          'data-testid': 'textfield-input',
        } as TextFieldProps);
        const input = component.getByTestId('textfield-input');
        expect(input).toBeDisabled();
      });
      it('should match the Disabled TextField snapshot', () => {
        const disabledTextField = createComponentJSX({
          disabled: true,
        } as TextFieldProps);
        const disabledTextFieldJSON = renderer
          .create(disabledTextField)
          .toJSON();
        expect(disabledTextFieldJSON).toMatchSnapshot();
      });
    });
  });
});
