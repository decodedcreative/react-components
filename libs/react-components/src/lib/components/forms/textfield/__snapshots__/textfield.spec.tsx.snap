// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textfield TextField - States TextField - States - Disabled should match the Disabled TextField snapshot 1`] = `
.c0 {
  width: 100%;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: #272727;
  font-family: 'Source Sans Pro',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  border: 2px solid #8e8e8e;
}

.c2:focus::not(.c1:focus-visible) {
  outline: 0;
  border-color: unset;
}

.c2:focus-visible {
  color: #515151;
  border-color: #006de3;
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
}

.c2:disabled {
  color: #8e8e8e;
  cursor: not-allowed;
}

<div
  className="c0"
>
  <input
    className="c1 c2"
    disabled={true}
    type="text"
  />
</div>
`;

exports[`Textfield TextField - States TextField - States - Error should match the TextField With Error snapshot 1`] = `
.c0 {
  width: 100%;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: #272727;
  font-family: 'Source Sans Pro',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  border: 2px solid #8e8e8e;
  border-color: #e90014;
  color: #e90014;
}

.c2:focus::not(.c1:focus-visible) {
  outline: 0;
  border-color: unset;
}

.c2:focus-visible {
  color: #515151;
  border-color: #006de3;
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
}

.c2:disabled {
  color: #8e8e8e;
  cursor: not-allowed;
}

<div
  className="c0"
>
  <input
    aria-invalid={true}
    className="c1 c2"
    type="text"
  />
</div>
`;

exports[`Textfield should match the default snapshot 1`] = `
.c0 {
  width: 100%;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: #272727;
  font-family: 'Source Sans Pro',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  border: 2px solid #8e8e8e;
}

.c2:focus::not(.c1:focus-visible) {
  outline: 0;
  border-color: unset;
}

.c2:focus-visible {
  color: #515151;
  border-color: #006de3;
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
}

.c2:disabled {
  color: #8e8e8e;
  cursor: not-allowed;
}

<div
  className="c0"
>
  <input
    className="c1 c2"
    type="text"
  />
</div>
`;
